adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -0,0 +1,384 @@
+            final String endpointName) {
+
+        final Future<JsonObject> tokenFuture = getRegistrationAssertion(context.getTenantId(), context.getDeviceId(),
+                null);
+        final Future<TenantObject> tenantConfigFuture = getTenantConfiguration(context.getTenantId());
+
+        CompositeFuture.all(tenantConfigFuture, tokenFuture, senderFuture).compose(ok -> {
+            final TenantObject tenantObject = tenantConfigFuture.result();
+            if (tenantObject.isAdapterEnabled(context.getTenantId())) {
+                final MessageSender sender = senderFuture.result();
+                final Message downstreamMessage = newMessage(context.getResourceIdentifier(),
+                        sender.isRegistrationAssertionRequired(),
+                        endpointName, context.getMessageContentType(), context.getMessagePayload(),
+                        tokenFuture.result(), null);
+
+                if (context.remotelySettled()) {
+                    // client uses AT_MOST_ONCE delivery semantics -> fire and forget
+                    return sender.send(downstreamMessage);
+                } else {
+                    // client uses AT_LEAST_ONCE delivery semantics
+                    return sender.sendAndWaitForOutcome(downstreamMessage);
+                }
+            } else {
+                // this adapter is not enabled for tenant
+                return Future.failedFuture(
+                        new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND,
+                                "This adapter is not enabled for tenant."));
+            }
+        }).compose(downstreamDelivery -> {
+            LOG.info("Successfully process message for Device [deviceId: {}] with Tenant [tenantId: {}]",
+                    context.getDeviceId(),
+                    context.getTenantId());
+            if (context.remotelySettled()) {
+                // client uses AT_MOST_ONCE delivery semantics
+                // accept & settle the message regardless of whether
+                // the downstream peer accepted/rejected the message.
+                context.accept();
+            } else {
+                // client uses AT_LEAST_ONCE delivery semantics
+                // forward disposition received from downstream peer back to the client device.
+                context.updateDelivery(downstreamDelivery.getRemoteState(),
+                        downstreamDelivery.remotelySettled());
+            }
+            return Future.<Void> succeededFuture();
+        }).recover(t -> {
+            LOG.info("Cannot process message for Device [deviceId: {}] with Tenant [tenantId: {}]",
+                    context.getDeviceId(),
+                    context.getTenantId());
+            if (!context.remotelySettled()) {
+                // client wants to be informed that the message cannot be processed.
+                context.released();
```

Comment:
shouldn't we reject the message in this case and include an error description?

Reply:
Oay! but maybe by releasing it, we are telling the client that the message is okay but the adapter cannot process it (because the sender queue was full, ...). The client can then try to redeliver the message.
IMHO, rejecting the message could signal to the client that something is actually wrong with the message. WDYT?

Reply:
Yes, I had the same thought. So it seems like we should distinguish between these two cases and then either release or reject the message, right? A ClientErrorException is used to indicate that the client has made an error (-> reject ?) while a ServerErrorException is used to indicate that there is a processing problem not caused by the client (-> release ?) ...

Reply:
> A ClientErrorException is used to indicate that the client has made an error (-> reject ?) while a ServerErrorException is used to indicate that there is a processing problem not caused by the client (-> release ?)

Yes, exactly! will update the PR